src/ptc/start.clj

```diff
@@ -1,122 +1,120 @@
+      (when (not (.hasMoreElements msg-enum))
+        (Thread/sleep 10000))
+      (.nextElement msg-enum))))
 
 (defn produce
   "Enqueue the TEXT with PROPERTIES map to JMS QUEUE through CONNECTION."
   [connection queue text properties]
   (letfn [(add-property [^TextMessage message k v]
-            (.setStringProperty message (name k) v))
-          (send [connection queue]
-            (with-open [session (create-session connection true)]
-              (let [queue (.createQueue session queue)
-                    message (.createTextMessage session text)]
-                (doseq [[k v] properties]
-                  (add-property message k v))
-                (with-open [producer (.createProducer session queue)]
-                  (.setDeliveryMode producer DeliveryMode/PERSISTENT)
-                  (.start connection)
-                  (.send producer message)
-                  (.commit session)))))]
-    (send connection queue)))
+            (.setStringProperty message (name k) v))]
+    (with-open [session (create-session-transacted connection)]
+      (let [queue   (.createQueue session queue)
+            message (.createTextMessage session text)]
+        (doseq [[k v] properties] (add-property message k v))
+        (with-open [producer (.createProducer session queue)]
+          (.setDeliveryMode producer DeliveryMode/PERSISTENT)
+          (.start connection)
+          (.send producer message)
+          (.commit session))))))
 
 (defn listen-and-consume-from-queue
   "Listen to QUEUE on CONNECTION for messages,
   and call (TASK! message) until it is false."
-  ([task! connection queue]
-   (loop [counter 0]
-     (if-let [peeked (peek-message connection queue)]
-       ; to avoid NPE on ednify
-       (let [peeked (jms/ednify peeked)]
-         (do (log/infof "Peeked message %s: %s" counter peeked)
-             (if (task! peeked)
-               (let [consumed (jms/ednify (consume connection queue))]
-                 (log/infof "Task complete, consumed message %s" counter)
-                 (if (not (misc/message-ids-equal? peeked consumed))
-                   (log/warnf
-                    (str/join \space ["Messages differ:"
-                                      (with-out-str (pprint (data/diff peeked consumed)))])))
-                 (recur (inc counter)))
-               (do
-                 (log/errorf
-                  (str/join
-                   \space ["Task returned nil/false,"
-                           "not consuming message %s and instead exiting"])
-                  counter)
-                 peeked))))
-       (recur counter))))
-  ([connection queue]
-   (let [ptc-bucket-name (or (System/getenv "PTC_BUCKET_NAME") "broad-gotc-dev-zero-test")
-         push-to (misc/gs-url ptc-bucket-name)
-         upload-sample! (fn [msg] (jms/handle-message push-to msg))]
-     (listen-and-consume-from-queue upload-sample! connection queue))))
+  [task! connection queue]
+  (loop [counter 0]
```

Comment:
There's still the problem that if this procedure throws, the counter will be re-set to zero.

Reply:
This will be addressed in https://broadinstitute.atlassian.net/browse/GH-893